### ISSUE TITLE
[pyright] reportUnboundVariable from CatchMissingPackages pattern

### ISSUE THREAD

**Issue by ringohoffman** [nerfstudio-project/nerfstudio#2770]
`nerfstudio` is using `pyright==1.1.331`, published 10/11/23, about 3 months ago. `pyright` publishes patches every week, and I think it is worth staying up to date with their latest releases.

Right now, when I run the latest `pyright` in Python 3.11 on `nerfstudio`, there are 103 errors. 7 of these errors are `reportUnboundVariable`:

```console
$ pip install pyright==1.1.347
$ pyright
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:68:38 - error: "DiffusionOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:103:38 - error: "DiffusionOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:129:38 - error: "DiffusionOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:205:38 - error: "DiffusionOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:208:38 - error: "IFOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/deepfloyd.py:212:41 - error: "IFOutputOrig" is possibly unbound (reportUnboundVariable)
  /home/matthew/Repositories/nerfstudio/nerfstudio/generative/stable_diffusion.py:77:33 - error: "DiffusionPipeline" is possibly unbound (reportUnboundVariable)
103 errors, 1 warning, 0 informations 
```
All of them are from the usage of `CatchMissingPackages`:

https://github.com/nerfstudio-project/nerfstudio/blob/3e2271a6563d76d06c0fc27c8e23440914e6020f/nerfstudio/generative/utils.py#L22-L35

In 2 files:

https://github.com/nerfstudio-project/nerfstudio/blob/3e2271a6563d76d06c0fc27c8e23440914e6020f/nerfstudio/generative/deepfloyd.py#L29-L35

https://github.com/nerfstudio-project/nerfstudio/blob/3e2271a6563d76d06c0fc27c8e23440914e6020f/nerfstudio/generative/stable_diffusion.py#L34-L38

Instances of `CatchMissingPackages` are used to mock members of modules in the `gen` `optional-depenencies` group that are not installed, and print a specialized error message when an attribute of one of these members are accessed. This has the effect of delaying the error message until `StableDiffusion.__init__()` or `DeepFloyd.__init__()` are called, instead of when the modules they are defined in are imported.

While I do appreciate a specialized error message that tells you exactly how you to solve it, I agree with `pyright` here that this pattern is a little confusing. Is there a specific reason that this error needs to be deferred until these constructors are called? Do we need to provide a specialized error message other than the `ModuleNotFoundError` that is automatically raised? Hopefully it shouldn't be too difficult for users to solve and can be avoided with proper documentation.

Are we open to refactoring this pattern in order to eliminate these `reportUnboundVariable` errors?

**Comment by brentyi** [comment]
I actually refactored these imports in #2763, will this still be an issue if that's merged?

**Comment by ringohoffman** [comment]
@brentyi on `brent/reduce_import_overhead` we are down to 83 errors and 0 `reportUnboundVariable` errors! Great work!

**Comment by ringohoffman** [comment]
Fixed by #2763
